distributed/shuffle/_worker_extension.py

```diff
@@ -351,158 +372,225 @@ def heartbeat(self) -> dict:
     async def shuffle_receive(
         self,
         shuffle_id: ShuffleId,
-        data: list[bytes],
+        run_id: int,
+        data: list[tuple[int, bytes]],
     ) -> None:
         """
         Handler: Receive an incoming shard of data from a peer worker.
         Using an unknown ``shuffle_id`` is an error.
         """
-        shuffle = await self._get_shuffle(shuffle_id)
+        shuffle = await self._get_shuffle_run(shuffle_id, run_id)
         await shuffle.receive(data)
 
-    async def shuffle_inputs_done(self, shuffle_id: ShuffleId) -> None:
+    async def shuffle_inputs_done(self, shuffle_id: ShuffleId, run_id: int) -> None:
         """
         Handler: Inform the extension that all input partitions have been handed off to extensions.
         Using an unknown ``shuffle_id`` is an error.
         """
         with log_errors():
-            shuffle = await self._get_shuffle(shuffle_id)
+            shuffle = await self._get_shuffle_run(shuffle_id, run_id)
             await shuffle.inputs_done()
-            if shuffle.done():
-                # If the shuffle has no output partitions, remove it now;
-                # `get_output_partition` will never be called.
-                # This happens when there are fewer output partitions than workers.
-                assert shuffle._disk_buffer.empty
-                logger.info(f"Shuffle inputs done {shuffle}")
-                await self._register_complete(shuffle)
-                del self.shuffles[shuffle_id]
-
-    async def shuffle_fail(self, shuffle_id: ShuffleId, message: str) -> None:
-        try:
-            shuffle = self.shuffles[shuffle_id]
-        except KeyError:
+
+    def shuffle_fail(self, shuffle_id: ShuffleId, run_id: int, message: str) -> None:
+        shuffle = self.shuffles.get(shuffle_id, None)
+        if shuffle is None or shuffle.run_id != run_id:
             return
+        self.shuffles.pop(shuffle_id)
         exception = RuntimeError(message)
         shuffle.fail(exception)
-        await shuffle.close()
-        del self.shuffles[shuffle_id]
+
+        async def _(extension: ShuffleWorkerExtension, shuffle: ShuffleRun) -> None:
+            await shuffle.close()
+            extension._runs.remove(shuffle)
+
+        self.worker._ongoing_background_tasks.call_soon(_, self, shuffle)
```

Comment:
why do we dispatch this to a background task? why not make shuffle_fail async?

Reply:
I can revisit whether we still need this. In some intermediate version, `shuffle_fail` needed to be synchronous to ensure ordering of operations. It might be that we solved this by propagating `ShuffleRun.run_id`.

Reply:
`shuffle_fail` must be synchronous to ensure that a shuffle run is always invalidated before the task of a new run starts computing. For this, we rely on the message ordering of the stream. However, we must also guarantee that the messages are processed in the order they were submitted to the stream. Since `handle_stream` hands coroutines off to background tasks instead of awaiting them, `shuffle_fail` must be synchronous to ensure that the shuffle is invalidated before we process the next message.

https://github.com/dask/distributed/blob/fa86772ff30dbe58922a05bb879b5cbbb74393fe/distributed/core.py#L898-L900

Note that this differs from the behavior of `handle_comm`, which awaits coroutines: https://github.com/dask/distributed/blob/fa86772ff30dbe58922a05bb879b5cbbb74393fe/distributed/core.py#L819-L820

Reply:
the above is no longer true. handlers are always executed in order since https://github.com/dask/distributed/pull/8059